src/Documentation/resources/Best_Practices.md

```diff
@@ -0,0 +1,421 @@
+
+    -   Fan-out:
+
+        -   `var tasks = new List\<Task\>();  
+            foreach(var grain in grains)  
+            { tasks.Add(grain.Foo()) }  
+            await Task.WhenAll(tasks);  
+            DoMoreWork();`
+
+**Implementation of Grains**:
+
+-   Never perform a thread-blocking operation within a grain. All operations other than local computations must be explicitly asynchronous.
+    
+    -   Examples: Synchronously waiting for an IO operation or a web service call, locking, running an excessive loop that is waiting for a condition, etc. 
+    
+-   When to use a [StatelessWorker]
+
+    -   Functional operations such as: decryption, decompression, and before
+        forwarding for processing
+
+    -   When only *local* grains are required in multiple activations
+
+    -   Example: Performs well with staged aggregation within local silo first
+
+-   Grains are non-reentrant by default
+
+    -   Deadlock can occur due to call cycles
+
+        -   Example: The grain calls itself
+
+    -   Timeouts are used to automatically break deadlocks
+
+    -   Attribute [Reentrant] can be used to allow the grain class reentrant
+
+    -   Reentrant is still single-threaded however, it may interleave (divide
+        processing/memory between tasks)
+
+    -   Handling interleaving increases risk by being error prone
+
+-   Inheritance
+
+    -   Inheriting grain interfaces is easy
+
+    -   Disambiguation may be needed to implement the same interface in multiple
+        grain glasses
+
+-   Generics are supported
+
+## Grain State Persistence
+
+Orleans’ grain state persistence APIs are designed to be easy-to-use and provide
+extensible storage functionality.
+
+-   Tutorial: *Needs to be created*
+
+**Overview**:
+
+-   Orleans.IGrainState is extended by a .NET interface which contains fields
+    that should be included in the grain’s persisted state.
+
+-   GrainBase\<T\> is extended by the grain class that adds a strongly typed
+    State property into the grain’s base class.
+
+-   The initial State.ReadStateAsync() automatically occurs prior to
+    ActiveAsync() has been called for a grain.
+
+-   When the grain’s state object’s data is changed, then the grain should call
+    State.WriteStateAsync()
+
+    -   Typically, grains call State.WriteStateAsync() at the end of grain
+        method to return the Write promise.
+
+    -   The Storage provider *could* try to batch Writes that may increase
+        efficiency, but behavioral contract and configurations are orthogonal
+        (independent) to the storage API used by the grain.
+
+    -   A **timer** is an alternative method to write updates periodically.
+
+        -   The timer allows the application to determine the amount of
+            “eventual consistency”/statelessness allowed.
+
+        -   Timing (immediate/none/minutes) can also be controlled as to when to
+            update.
+
+    -   Each grain class can only be associated with one storage provider.
```

Comment:
This part wouldn't be needed for `IPersistentState<TState>` or can be rephrased for it instead.

Reply:
Updated to, "PersistetState classes, like other grain classes, can only be associated with one storage provider." Thanks again!